### PR TITLE
Fix some valid curved label placements are rejected 

### PR DESCRIPTION
**Pull request by nyalldawson**
This happens predominantly in maps with geographic CRS, but is also seen in large scale projected maps. The calculation for
line circle intersections used to determine curved label character placement becomes numerically unstable when the map extent width/height is small, leading to incorrect angular calculations for individual labels and ultimately the candidate being rejected as violating the max in/out character angle constraints.

Shown below is the difference of labeling results using the exact same conditions prior to the fix and after:

Peek 2021-04-05 12-02

### CODE REVIEW

File: src/core/pal/feature.cpp [hunk 1]
```diff
@@ -1223,7 +1223,7 @@ std::size_t FeaturePart::createCandidatesAlongLineNearMidpoint( std::vector< std
 }
 
 
-std::unique_ptr< LabelPosition > FeaturePart::curvedPlacementAtOffset( PointSet *path_positions, double *path_distances, int &orientation, const double offsetAlongLine, bool &reversed, bool &flip, bool applyAngleConstraints )
+std::unique_ptr< LabelPosition > FeaturePart::curvedPlacementAtOffset( PointSet *path_positions, const std::vector< double> &path_distances, int &orientation, const double offsetAlongLine, bool &reversed, bool &flip, bool applyAngleConstraints )
```
Comment (nirvn) [inline src/core/pal/feature.cpp:1226]: Seize the opportunity to rename the path_distances variable to pathDistances?

File: src/core/pal/feature.cpp [hunk 2]
```diff
@@ -1240,7 +1240,7 @@ std::unique_ptr< LabelPosition > FeaturePart::curvedPlacementAtOffset( PointSet
 
   LabelInfo *li = mLF->curvedLabelInfo();
 
-  double string_height = li->label_height;
+  double string_height = li->characterHeight;
```
Comment (nirvn) [inline src/core/pal/feature.cpp:1243]: Also here, stringHeight? I'm mentioning it because I see you've (nicely) done so in other parts.

**Review by nirvn — APPROVED** [review]
A few suggestions, which aren't meant to block this PR if you don't want to do.

Can't claim I'm on top of this part of the code, but can't see anything wrong with your changes :smile:

File: src/core/pal/feature.h [hunk 3]
```diff
@@ -265,16 +265,16 @@ namespace pal
 
       /**
        * Returns the label position for a curved label at a specific offset along a path.
-       * \param path_positions line path to place label on
-       * \param path_distances array of distances to each segment on path
+       * \param mapShape line path to place label on
```
Comment (nirvn) [inline src/core/pal/feature.h:268]: @nyalldawson , mapShape or pathShape?

Thanks for the extra cleanup.

Comment (nyalldawson) [reply src/core/pal/feature.h:268]: I've gone with "mapShape" to match all the other methods in this class

Comment (nirvn) [reply src/core/pal/feature.h:268]: You undoubtedly know best :)

**Pull request merged**